### PR TITLE
feat: add message list

### PR DESCRIPTION
**Pull request by bhavya2611**
## What does this PR do?

Adds a message list component and a message composer component. Please note, the height of the containers is something to be fixed and will be done later, minor styling issues might also be present which are not given top priority yet.

## Type of change

- [ ] Enhancement (non-breaking small changes to existing functionality)

## How should this be tested?

- [ ] See if the messages get loaded correctly for each user

## Screenshots

image

image

### CODE REVIEW

File: src/components/Shared/MessageComposer.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { Button } from '@components/UI/Button';
+import { Input } from '@components/UI/Input';
+import { ArrowRightIcon } from '@heroicons/react/outline';
+import { FC, useState } from 'react';
+
+interface Props {
+  // eslint-disable-next-line no-unused-vars
```
Comment (bigint) [inline src/components/Shared/MessageComposer.tsx:7]: **suggestion:** You can remove this comment 😊

File: src/components/Shared/MessageComposer.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      <Input
+        value={message}
+        onChange={(e) => setMessage(e.target.value)}
+        type="text"
+        placeholder="Type Something"
+      />
+      <Button
+        onClick={handleSend}
+        className="text-md ml-2 !px-4 !py-1.5 min-w-max"
+        variant="success"
```
Comment (bigint) [inline src/components/Shared/MessageComposer.tsx:30]: **suggestion:** we can make this primary 🟣

File: src/components/Shared/MessagesList.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    : '';
+
+const isOnSameDay = (d1?: Date, d2?: Date): boolean => {
+  return d1?.toDateString() === d2?.toDateString();
+};
+
+const formatDate = (d?: Date) =>
+  d?.toLocaleString('en-US', { year: 'numeric', month: 'long', day: 'numeric' });
+
+const MessageTile = ({ message }: MessageTileProps): JSX.Element => {
```
Comment (bigint) [inline src/components/Shared/MessagesList.tsx:34]: **suggestion:** you can use something like this

```
const MessageTile: FC<MessageTileProps> = ({ message }) => {
```

https://github.com/lensterxyz/lenster/blob/main/src/components/Home/Hero.tsx#L4

File: src/components/Shared/MessagesList.tsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
+import getAvatar from '@lib/getAvatar';
+import { Message } from '@xmtp/xmtp-js';
+import React, { ReactNode } from 'react';
+import { useAppStore } from 'src/store/app';
+
+export type MessageListProps = {
+  messages: Message[];
+};
+
+type MessageTileProps = {
```
Comment (bigint) [inline src/components/Shared/MessagesList.tsx:10]: **suggestion:** better to use interface everywhere 🙏🏼

File: src/components/Shared/MessagesList.tsx [hunk 5]
```diff
@@ -0,0 +1,123 @@
+import getAvatar from '@lib/getAvatar';
+import { Message } from '@xmtp/xmtp-js';
+import React, { ReactNode } from 'react';
+import { useAppStore } from 'src/store/app';
+
+export type MessageListProps = {
```
Comment (bigint) [inline src/components/Shared/MessagesList.tsx:6]: **question:** why export?

File: src/components/Shared/MessagesList.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+const MessageTile = ({ message }: MessageTileProps): JSX.Element => {
+  const currentProfile = useAppStore((state) => state.currentProfile);
+  const address = currentProfile?.ownedBy;
+
+  return (
+    <div
+      className={`flex flex-col ${
+        address === message.senderAddress ? 'items-end' : 'items-start'
+      } mx-auto mb-4`}
```
Comment (bigint) [inline src/components/Shared/MessagesList.tsx:42]: **nitpick:** using `clsx` will be easy to read

example: https://github.com/lensterxyz/lenster/blob/ac21706ff46bcd1139222ffcb36dd7bc0098781d/src/components/UI/Toggle.tsx#L17

File: src/components/Shared/MessagesList.tsx [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+const formatTime = (d: Date | undefined): string =>
+  d
+    ? `${d.toLocaleTimeString(undefined, {
+        hour12: true,
+        hour: 'numeric',
+        minute: '2-digit'
+      })} - ${d.toLocaleDateString()}`
+    : '';
```
Comment (bigint) [inline src/components/Shared/MessagesList.tsx:25]: **nitpick:** we are using https://day.js.org/ for formating dates 🙏🏼

File: src/components/Shared/MessagesList.tsx [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        )}
+        <div
+          className={`px-4 py-2 rounded-lg ${
+            address === message.senderAddress ? 'bg-brand-500' : 'bg-gray-100'
+          } `}
+        >
+          <span
+            className={`block text-md font-normal ${
+              address === message.senderAddress ? 'text-white' : 'text-black'
+            } `}
```
Comment (bigint) [inline src/components/Shared/MessagesList.tsx:60]: **suggestion:** `clsx` here too

File: src/components/Shared/MessagesList.tsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      <div className={`${address !== message.senderAddress ? 'ml-12' : ''}`}>
+        <span className="text-xs font-normal place-self-end text-gray-400 uppercase">
+          {formatTime(message.sent)}
+        </span>
+      </div>
+    </div>
+  );
+};
+
+const DateDividerBorder: React.FC<Props> = ({ children }) => (
```
Comment (bigint) [inline src/components/Shared/MessagesList.tsx:75]: **suggestion:** just import `FC` instead of using `React.FC`

**Review by bigint — CHANGES_REQUESTED** [review]
@bhavya2611 left some comments 🙇🏼

**Comment by bigint** [comment]
image

Seems like `useStreamMessages` needs 2 arguments 🤔

**Review by bigint — APPROVED** [review]
Pushed some changes 🚆 

**LGTM** :shipit: 

Rad LGTM

**Comment by bigint** [comment]
@elisealix22 @neekolas can you review this as well 🙏🏼

File: src/components/Messages/Message.tsx [hunk 10]
```diff
@@ -11,14 +14,17 @@ import { useAppStore } from 'src/store/app';
...
 
 const Message: FC = () => {
-  const router = useRouter();
-  const address = router.query.address as string;
-  const messageState = useMessageStore((state) => state);
-  const { messages, conversations } = messageState;
+  const { push, query } = useRouter();
+  const address = query.address as string;
   const currentProfile = useAppStore((state) => state.currentProfile);
+  const messageState = useMessageStore((state) => state);
```
Comment (neekolas) [inline src/components/Messages/Message.tsx:20]: I'm not a Zustand expert, so correct me if I'm wrong, but won't returning the complete state cause the component to rerender any time anything in the state changes?

It seems like we want to only select the items we care about, and use `zustand/shallow` when selecting multiple items.

https://docs.pmnd.rs/zustand/recipes/recipes

File: src/components/utils/hooks/useGetMessages.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+const useGetMessages = (conversation?: Conversation) => {
+  const messageState = useMessageStore((state) => state);
+  const { messages, setMessages } = messageState;
+
+  useEffect(() => {
+    if (!conversation) {
+      return;
+    }
+    const loadMessages = async () => {
+      const newMessages = await conversation.messages();
```
Comment (neekolas) [inline src/components/utils/hooks/useGetMessages.tsx:14]: Don't worry @bigint, we are going to add pagination to make this more performant. Just prioritizing getting the UI in place first.

File: src/components/Messages/Message.tsx [hunk 12]
```diff
@@ -11,14 +14,17 @@ import { useAppStore } from 'src/store/app';
...
+  const { conversations } = messageState;
+  const selectedConversation = conversations.get(address);
+  const { messages } = useGetMessages(selectedConversation);
+  const { sendMessage } = useStreamMessages(selectedConversation);
```
Comment (neekolas) [inline src/components/Messages/Message.tsx:24]: Not quite sure why `sendMessage` is included in `useStreamMessages`. Feels unrelated to me

Comment (neekolas) [reply src/components/Messages/Message.tsx:24]: Also, I noticed that `sendMessage` isn't memoized with `useCallback` when it should be

File: src/components/Messages/Message.tsx (hunk 10 above)
Comment (bigint) [reply src/components/Messages/Message.tsx:20]: Yes it's always good to select each items individually 🙇🏼 

Example: https://github.com/lensterxyz/lenster/blob/main/src/components/Layout.tsx#L35-L41

File: src/components/utils/hooks/useGetMessages.tsx (hunk 11 above)
Comment (bigint) [reply src/components/utils/hooks/useGetMessages.tsx:14]: Cool ⚡
